content/2017/03/16/serializing-things.rst

```diff
@@ -0,0 +1,210 @@
+
+.. code-block:: python
+
+    from django.apps import apps
+
+    @task()
+    def some_task(model_name, model_id):
+        Model = apps.get_model('django_app_name.{}'.format(model_name))
+        obj = Model.objects.get(pk=model_id)
+        # Operate on obj
+
+Note the crucial piece here: ``django.apps.apps.get_model``. It takes a model
+identifier, which is ``django_app_name.ModelName``. The ``django_app_name`` is
+the last dot-separated part of whatever you put in ``INSTALLED_APPS``. The
+``ModelName`` is the name of the class in the ``models`` module in that app.
+
+For added delight here, you can even get the model name automatically in a
+mixin to your models:
+
+.. code-block:: python
+
+    class SomeMixin:
+        # Assuming that you want to trigger the task on save:
+        def save(self, *args, **kwargs):
+            ret = super().save(*args, **kwargs)
+            some_task.apply_async((
+                self.__class__.__name__,
+                self.pk,
+            ))
+            return ret
+
+As a final word to the wise, it's worth noting that this entire
+database-mediated approach opens you up to certain timing risks. Data can skew,
+and you expose yourself to potential race conditions. Sometimes that's not an
+issue, and sometimes it's just an acceptable cost. But in any case, it's worth
+keeping in mind.
+
+
+But what if I need something that's not a model?
+------------------------------------------------
+
+Perhaps you have business-logic class instances which are never stored in the
+database.[#]_ If you can't, won't, or don't want to use the DB as a persistent
+store for your data, which you then inflate into a full object, there are
+some other ways to pass objects through the task-broker bottleneck.
+
+They all boil down to separating the primitive data from the methods and logic.
+Think of it like passing the *record* or *struct* through, not the whole class.
+
+So if that's the goal, you could make custom JSON encoders and decoders that
+know how to traverse your classes. But that's a pain. Let's see if we can write
+as little code not related to our actual business logic as possible.
+
+One approach I like is to use the `attrs
+<https://attrs.readthedocs.io/en/stable/>`_ library. It lets you define your
+business logic class like so:
+
+.. code-block:: python
+
+    import attr
+
+    @attr.s
+    class SomeClass(object):
+        foo = attr.ib()
+        bar = attr.ib()
+
+        def some_method(self):
+            pass
+
+And then you can serialize an instance easily:
+
+.. code-block:: python
+
+    import attr
+    inst = SomeClass(foo={'hi': 'there'}, bar=SomeClass(foo=1, bar=False))
+    attr.asdict(inst)
+    # {'foo': {'bar': False, 'foo': 1}, 'foo': {'hi': 'there'}}
+
+And just as importantly, you can pass that serialized data to the task, and
+inflate it like:
+
+.. code-block:: python
+
+    def some_task(some_class):
+        inst = SomeClass(**some_class)
+
+Try it out! It's very fun.
```

Comment:
You can grab the closing Slack/Twitter CTA thing on one of the other posts and re-word it a bit if you'd like?